azuki-script-generation/src/main/kotlin/com/anaplan/engineering/azuki/script/generation/ScriptGenerationService.kt

```diff
@@ -0,0 +1,272 @@
+package com.anaplan.engineering.azuki.script.generation
+
+import com.anaplan.engineering.azuki.core.system.*
+import com.anaplan.engineering.azuki.declaration.*
+
+/**
+ * The main endpoint for script generation tasks.
+ */
+class ScriptGenerationService<
+    // mandatory factories
+    out AF : ActionFactory,
+    out CF : CheckFactory,
+    // optional factories (these require calling 'withXYZFactory')
+    out QF : QueryFactory,
+    out AGF : ActionGeneratorFactory,
+    // mandatory non-factory parameters
+    DS : DeclarationState,
+    // optional non-factory parameters (these require calling 'withXYZ')
+    E : ScriptGenerationEnvironment,
+    > private constructor(
+    // mandatory parameters
+    internal val actionFactory: AF,
+    internal val checkFactory: CF,
+    internal val declarationStateFactory: DeclarationStateFactory<DS>,
+    // optional parameters
+    internal val environmentFactory: ScriptGenerationEnvironmentFactory<E>,
+    internal val queryQueryFactory: QF,
+    internal val verifyQueryFactory: QF,
+    internal val actionGeneratorFactory: AGF,
+) {
+
+    /**
+     * Provides ready-made usage patterns for script generation of a whole scenario, from a whole scenario.
+     */
+    val scenario = ScriptGenerationScenarioPatterns(this)
+
+    /**
+     * Adds an environment factory to this service, changing the type of accepted scenarios accordingly.
+     */
+    fun <N : ScriptGenerationEnvironment> withEnvironmentFactory(new: ScriptGenerationEnvironmentFactory<N>) =
+        ScriptGenerationService(
+            actionFactory,
+            checkFactory,
+            declarationStateFactory,
+            environmentFactory = new,
+            queryQueryFactory,
+            verifyQueryFactory,
+            actionGeneratorFactory,
+        )
+
+    /**
+     * Adds query factories to this service, changing the type of accepted scenarios accordingly.
+     */
+    fun <N : QueryFactory> withQueryFactories(query: N, verify: N) = ScriptGenerationService(
+        actionFactory,
+        checkFactory,
+        declarationStateFactory,
+        environmentFactory,
+        queryQueryFactory = query,
+        verifyQueryFactory = verify,
+        actionGeneratorFactory,
+    )
+
+    /**
+     * Adds an action generation factory to this service, changing the type of accepted scenarios accordingly.
+     */
+    fun <N : ActionGeneratorFactory> withActionGeneratorFactory(new: N) = ScriptGenerationService(
+        actionFactory,
+        checkFactory,
+        declarationStateFactory,
+        environmentFactory,
+        queryQueryFactory,
+        verifyQueryFactory,
+        actionGeneratorFactory = new,
+    )
```

Comment:
I wonder if we need a more obvious stop between these methods (which are _building_ a `ScriptGenerationService` and the next ones (which are _consuming_ it).  It seemed a bit pointless to introduce a separate builder type (`ScriptGenerationService` _is_ a builder!), but the two do run into each other awkwardly here.